apex/contrib/csrc/peer_memory/peer_memory_cuda.cu

```diff
@@ -0,0 +1,481 @@
+        for (auto it = shape.rbegin();  it != shape.rend();  ++it)
+        {
+	    strides[--idx] = size;
+	    size *= *it;
+        }
+    }
+    size *= sizeof(T);
+    // TODO: Implement dynamic reuse of pooled peer memory.
+    // We provide no deleter function because all peer memory allocations are static in this implementation.
+    return torch::from_blob((void*)raw_ptr, shape, strides, 0L, options);
+}
+
+void tensor_shape(at::Tensor t, bool explicit_nhwc, int& N, int& C, int& H, int& W)
+{
+    if (t.dim() == 3) {
+	N = 1;
+        if (explicit_nhwc) {
+            C = t.size(2);
+            H = t.size(0);
+            W = t.size(1);
+        } else {
+	    C = t.size(0);
+    	    H = t.size(1);
+    	    W = t.size(2);
+        }
+    } else if (t.dim() == 4) {
+        if (explicit_nhwc) {
+            N = t.size(0);
+            C = t.size(3);
+            H = t.size(1);
+            W = t.size(2);
+        } else {
+            N = t.size(0);
+            C = t.size(1);
+            H = t.size(2);
+            W = t.size(3);
+        }
+    } else {
+        printf("%s;%d - t.dim() must be either 3 or 4 (was %d)\n",__FILE__,__LINE__,t.dim());
+        assert(t.dim() == 3 || t.dim() == 4);
+    }
+}
+
+void tensor_strides(at::Tensor t, bool explicit_nhwc, int& stride_N, int& stride_C, int& stride_H, int& stride_W)
+{
+    if (t.dim() == 3) {
+        if (explicit_nhwc) {
+            stride_C = t.stride(2);
+            stride_H = t.stride(0);
+            stride_W = t.stride(1);
+        } else {
+	    stride_C = t.stride(0);
+    	    stride_H = t.stride(1);
+    	    stride_W = t.stride(2);
+        }
+        stride_N = t.size(0)*t.size(1)*t.size(2);
+    } else if (t.dim() == 4) {
+        if (explicit_nhwc) {
+            stride_N = t.stride(0);
+            stride_C = t.stride(3);
+            stride_H = t.stride(1);
+            stride_W = t.stride(2);
+        } else {
+            stride_N = t.stride(0);
+            stride_C = t.stride(1);
+            stride_H = t.stride(2);
+            stride_W = t.stride(3);
+        }
+    } else {
+        printf("%s;%d - t.dim() must be either 3 or 4 (was %d)\n",__FILE__,__LINE__,t.dim());
+        assert(t.dim() == 3 || t.dim() == 4);
+    }
+}
+
+template<class T, bool is_HWC>
+__device__ void strided_copy_kernel(
+	T* dst, const int dst_stride_C, const int dst_stride_H, const int dst_stride_W, 
+	const T* src, const int src_stride_C, const int src_stride_H, const int src_stride_W, 
+	const int NC, const int NH, const int NW
+	)
+{
+    size_t tot_num_threads = gridDim.x * blockDim.x;
+    size_t thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+    const size_t count = NC*NH*NW;
+    for (size_t i = thread_id;  i < count;  i += tot_num_threads)
+    {
+	size_t c,h,w;
+	if (is_HWC) {
+	    c = i % NC;
+	    w = i / NC;
+	    h = w / NW;
+	    w = w % NW;
+	}
+	else {
+	    w = i % NW;
+	    h = i / NW;
+	    c = h / NH;
+            h = h % NH;
```

Comment:
Maybe consider using the int fastdiv trick here to avoid repeated integer divide/modulus by the same value?

Reply:
I hadn't heard of the fastdiv trick, but it makes perfect sense. How do I use this in my code? Maxim Milakov has an implementation on github https://github.com/milakov/int_fastdiv. Can I just use his int_fastdiv.h header file?